### PR TITLE
Cache/dashboard

### PR DESCRIPTION
**Pull request by maruf0011**
Closes: #25 #24

### CODE REVIEW

File: src/collector/collector.ts [hunk 1]
```diff
@@ -50,6 +52,10 @@ const jobs = [
   {
     method: proposalCollector.run.bind(proposalCollector),
     cron: '30 * * * * *'
+  },
+  {
+    method: dashboardCollector.run.bind(dashboardCollector),
+    cron: '5 1 * * * *'
```
Comment (hanjukim) [inline src/collector/collector.ts:58]: why is this 5 1 instead of 0 0?

File: src/collector/dashboard/cacheDashboard.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import { getAccountCountByDay } from './accountGrowth'
+import { getBlockRewardsByDay } from './blockReward'
+import { getTxVolumeByDay } from './txVolume'
+import { getDateFromDateTime } from './helpers'
+import config from 'config'
+import { collectorLogger as logger } from 'lib/logger'
+
+const PREVIOUS_DAYS_TO_CALCULATE = 5
+
+export async function collectDashboard() {
```
Comment (hanjukim) [inline src/collector/dashboard/cacheDashboard.ts:15]: file name should be renamed to collectDashboard

File: src/collector/general.ts [hunk 3]
```diff
@@ -31,6 +31,22 @@ export async function saveGeneral() {
   const stakingPool = await lcd.getStakingPool()
   const taxProceeds = await lcd.getTaxProceeds()
   const seigniorageProceeds = await lcd.getSeigniorageProceeds()
+  const communityPoolList = await lcd.getCommunityPool()
+
+  const communityPool: DenomMap = communityPoolList.reduce((acc, { denom, amount }) => {
```
Comment (hanjukim) [inline src/collector/general.ts:36]: I think we can combine above promises to Promise.all

File: src/scripts/cacheDashboard.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
+import { getRepository } from 'typeorm'
```
Comment (hanjukim) [inline src/scripts/cacheDashboard.ts:1]: How about we rename this to populateDashboard

File: src/collector/collector.ts (hunk 1 above)
Comment (maruf0011) [reply src/collector/collector.ts:58]: just to make sure it wont miss any data at 0 0 as it uses others collector data. We can move use 0 30 too.

Comment (hanjukim) [reply src/collector/collector.ts:58]: ok than let's change it to 0 1

File: src/collector/dashboard/accountGrowth.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+async function getDailyActiveAccount(daysBefore?: number): Promise<{ date: string; active_account_count: number }[]> {
+  const queryBuilder = getRepository(AccountTxEntity)
+    .createQueryBuilder()
+    .select(convertDbTimestampToDate('timestamp'), 'date')
+    .addSelect('count(distinct account)', 'active_account_count')
+    .groupBy('date')
+    .orderBy('date', 'ASC')
+    .where('timestamp < :today ', { today: startOfToday() })
+
```
Comment (hanjukim) [inline src/collector/dashboard/accountGrowth.ts:41]: Can we just rely on raw query for counting unique count? the query time is crucial here. I have experienced that above query was significantly faster than count(distinct account).

```
fcd=> EXPLAIN ANALYZE SELECT COUNT(account) FROM (SELECT DISTINCT account FROM account_tx WHERE timestamp >='2020-04-18' AND timestamp < '2020-05-17') a;

 Aggregate  (cost=1910387.93..1910387.94 rows=1 width=8) (actual time=3906.017..3906.017 rows=1 loops=1)
   ->  HashAggregate  (cost=1909951.07..1910145.23 rows=19416 width=45) (actual time=3798.338..3881.513 rows=416933 loops=1)
         Group Key: account_tx.account
         ->  Bitmap Heap Scan on account_tx  (cost=133587.14..1894239.23 rows=6284739 width=45) (actual time=424.311..1854.403 rows=9014411 loops=1)
               Recheck Cond: (("timestamp" >= '2020-04-18 00:00:00'::timestamp without time zone) AND ("timestamp" < '2020-05-17 00:00:00'::timestamp without time zone))
               Rows Removed by Index Recheck: 31
               Heap Blocks: exact=47284 lossy=164614
               ->  Bitmap Index Scan on account_tx_timestamp  (cost=0.00..132015.96 rows=6284739 width=0) (actual time=409.005..409.005 rows=9016374 loops=1)
                     Index Cond: (("timestamp" >= '2020-04-18 00:00:00'::timestamp without time zone) AND ("timestamp" < '2020-05-17 00:00:00'::timestamp without time zone))
 Planning time: 1.896 ms
 Execution time: 3909.978 ms
```
```
fcd=> EXPLAIN ANALYZE SELECT COUNT(DISTINCT account) FROM account_tx WHERE timestamp >='2020-04-18' AND timestamp < '2020-05-17';

 Aggregate  (cost=1909956.07..1909956.08 rows=1 width=8) (actual time=42255.983..42255.984 rows=1 loops=1)
   ->  Bitmap Heap Scan on account_tx  (cost=133587.14..1894244.23 rows=6284739 width=45) (actual time=413.322..1901.046 rows=9014411 loops=1)
         Recheck Cond: (("timestamp" >= '2020-04-18 00:00:00'::timestamp without time zone) AND ("timestamp" < '2020-05-17 00:00:00'::timestamp without time zone))
         Rows Removed by Index Recheck: 31
         Heap Blocks: exact=47284 lossy=164614
         ->  Bitmap Index Scan on account_tx_timestamp  (cost=0.00..132015.96 rows=6284739 width=0) (actual time=403.426..403.426 rows=9016374 loops=1)
               Index Cond: (("timestamp" >= '2020-04-18 00:00:00'::timestamp without time zone) AND ("timestamp" < '2020-05-17 00:00:00'::timestamp without time zone))
 Planning time: 0.128 ms
 Execution time: 42256.945 ms
````

**Comment by hanjukim** [comment]
I think we can remove dashboardRawQuery now.

**Comment by maruf0011** [comment]
dashboardRawQuery is still in use by priceHistory.

File: src/service/dashboard/dashboardHistory.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import config from 'config'
+import { DashboardEntity } from 'orm'
+
+export async function getDashboardHistory(daysBefore?: number): Promise<DashboardEntity[]> {
+  const whereClause = {
+    chainId: config.CHAIN_ID
+  }
+
+  if (daysBefore) {
+    whereClause['timestamp'] = MoreThanOrEqual(subDays(startOfToday(), daysBefore))
```
Comment (hanjukim) [inline src/service/dashboard/dashboardHistory.ts:13]: I found these tests are failing
```
    ✕ Test get account growth count filter (24 ms)
    ✕ Test get tx volumes with count filter (18 ms)
    ✕ Test get block rewards with count filter (19 ms)
```
I can fix it by adding 1 to daysBefore variable. Please confirm.

Comment (maruf0011) [reply src/service/dashboard/dashboardHistory.ts:13]: it's failed because data didn't generated in test db due to error in dashboardCollector. Now fixed.

**Pull request merged**